File: app/src/main/java/com/worldventures/dreamtrips/modules/friends/service/command/ActOnFriendRequestCommand.java
package com.worldventures.dreamtrips.modules.friends.service.command;

import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.api.friends.AnswerFriendRequestsHttpAction;
import com.worldventures.dreamtrips.api.friends.model.FriendRequestParams;
import com.worldventures.dreamtrips.core.api.action.CommandWithError;
import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.modules.common.model.User;

import javax.inject.Inject;

import io.techery.janet.Janet;
import io.techery.janet.command.annotations.CommandAction;

public abstract class ActOnFriendRequestCommand extends CommandWithError<User> implements InjectableAction {

   protected User user;

   @Inject Janet janet;

   public ActOnFriendRequestCommand(User user) {
      this.user = user;
   }

   @Override
   protected void run(CommandCallback<User> callback) throws Throwable {
      janet.createPipe(AnswerFriendRequestsHttpAction.class)
            .createObservableResult(new AnswerFriendRequestsHttpAction(getRequestParams()))
            .subscribe(action -> {
               callback.onSuccess(user);
            }, callback::onFail);
   }

   abstract FriendRequestParams getRequestParams();

   @CommandAction
   public static final class Accept extends ActOnFriendRequestCommand {

      private String circleId;

      public Accept(User user, String circleId) {
         super(user);
         this.circleId = circleId;
      }

      @Override
      public int getFallbackErrorMessage() {
         return R.string.error_fail_to_accept_friend_request;
      }

      @Override
      FriendRequestParams getRequestParams() {
         return FriendRequestParams.confirm(user.getId(), circleId);
      }
   }

   @CommandAction
   public static final class Reject extends ActOnFriendRequestCommand {

      public Reject(User user) {
         super(user);
      }

      @Override
      public int getFallbackErrorMessage() {
         return R.string.error_fail_to_reject_friend_request;
      }

      @Override
      FriendRequestParams getRequestParams() {
         return FriendRequestParams.reject(user.getId());
      }
   }

}

File: app/src/main/java/com/worldventures/dreamtrips/social/service/friends/interactor/command/ActOnFriendRequestCommand.kt
package com.worldventures.dreamtrips.social.service.friends.interactor.command

import com.worldventures.core.janet.CommandWithError
import com.worldventures.core.model.User
import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.api.friends.AnswerFriendRequestsHttpAction
import com.worldventures.dreamtrips.api.friends.model.FriendRequestParams
import com.worldventures.janet.injection.InjectableAction
import io.techery.janet.Janet
import io.techery.janet.command.annotations.CommandAction
import javax.inject.Inject

@CommandAction
abstract class ActOnFriendRequestCommand(val user: User) : CommandWithError<User>(), InjectableAction {

   @field:Inject lateinit var janet: Janet

   @Throws(Throwable::class)
   override fun run(callback: CommandCallback<User>?) {
      janet.createPipe(AnswerFriendRequestsHttpAction::class.java)
            .createObservableResult(AnswerFriendRequestsHttpAction(getRequestParams()))
            .subscribe({ callback?.onSuccess(user) }, { callback?.onFail(it) })
   }

   abstract fun getRequestParams(): FriendRequestParams

   @CommandAction
   class Accept(user: User, val circleId: String) : ActOnFriendRequestCommand(user) {
      override fun getRequestParams(): FriendRequestParams {
         return FriendRequestParams.confirm(user.id, circleId)
      }

      override fun getFallbackErrorMessage(): Int = R.string.error_fail_to_accept_friend_request
   }

   @CommandAction
   class Reject(user: User) : ActOnFriendRequestCommand(user) {
      override fun getRequestParams(): FriendRequestParams {
         return FriendRequestParams.reject(user.id)
      }

      override fun getFallbackErrorMessage(): Int = R.string.error_fail_to_reject_friend_request
   }
}
